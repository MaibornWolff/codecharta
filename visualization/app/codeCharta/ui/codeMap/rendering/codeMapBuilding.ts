import * as THREE from "three";
import {MapColors} from "./renderSettings";
import {node} from "./node";

export class codeMapBuilding {
    public id : number;
    public boundingBox : THREE.Box3;
    public color : number;
    public node : node;
    public parent?: codeMapBuilding;
    public name?: string;

    constructor(id : number, box : THREE.Box3, node : node, color? : number)
    {
        this.id = id;
        this.boundingBox = box;
        this.color = color ? color : MapColors.defaultC;
        this.node = node;
    }
}