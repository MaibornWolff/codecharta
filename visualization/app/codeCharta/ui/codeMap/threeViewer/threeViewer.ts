import {ThreeViewerService} from "./threeViewerService";
import {ThreeSceneService} from "./threeSceneService";
import {ThreeCameraService} from "./threeCameraService";
import {ThreeOrbitControlsService} from "./threeOrbitControlsService";
import {ThreeRendererService} from "./threeRendererService";
import {ThreeUpdateCycleService} from "./threeUpdateCycleService";
import "../../../core/settings/settings.module";

import angular from "angular";

angular.module("app.codeCharta.ui.codeMap.threeViewer", ["app.codeCharta.core.settings"])
    .service(
        ThreeViewerService.SELECTOR,
        ThreeViewerService
    )
    .service(
        ThreeUpdateCycleService.SELECTOR,
        ThreeUpdateCycleService
    )
    .service(
        ThreeSceneService.SELECTOR,
        ThreeSceneService
    )
    .service(
        ThreeRendererService.SELECTOR,
        ThreeRendererService
    )
    .service(
        ThreeOrbitControlsService.SELECTOR,
        ThreeOrbitControlsService
    )
    .service(
        ThreeCameraService.SELECTOR,
        ThreeCameraService
    );


