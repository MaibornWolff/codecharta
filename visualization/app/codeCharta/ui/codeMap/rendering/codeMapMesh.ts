import * as THREE from "three";

import {codeMapShaderStrings} from "./codeMapShaderStrings";
import {geometryGenerator, buildResult} from "./geometryGenerator";
import {codeMapGeometricDescription} from "./codeMapGeometricDescription";
import {intersectionResult} from "./codeMapGeometricDescription";
import {codeMapBuilding} from "./codeMapBuilding";
import {renderingUtil} from "./renderingUtil";
import {node} from "./node";

import {MapColors} from "./renderSettings";
import {renderSettings} from "./renderSettings";

interface threeUniform {
    type : string;
    value : any;
}

interface codeMapLightingParams {
    numHighlights : threeUniform;
    numSelections : threeUniform;
    highlightColor : threeUniform;
    highlightedIndices : threeUniform;
    selectedColor : threeUniform;
    selectedIndices : threeUniform;
    emissive : threeUniform;
    deltaColorPositive : threeUniform;
    deltaColorNegative : threeUniform;
}

export interface mousePos {
    x : number;
    y : number;
}

export class CodeMapMesh {
    private threeMesh : THREE.Mesh;
    private material : THREE.ShaderMaterial;
    private geomGen : geometryGenerator;
    private mapGeomDesc : codeMapGeometricDescription;

    private nodes : node[];

    private currentlyHighlighted : codeMapBuilding[] | null;
    private currentlySelected : codeMapBuilding[] | null;

    public settings : renderSettings;

    private lightingParams : codeMapLightingParams = {
        numHighlights : {type : "f", value : 0.0},
        highlightColor : {type : "v3", value : renderingUtil.colorToVec3(0x666666)},
        highlightedIndices : {type : "fv1", value : []},

        numSelections : {type : "f", value : 0.0},
        selectedColor : {type : "f", value : renderingUtil.colorToVec3(MapColors.selected)},
        selectedIndices : {type : "fv1", value : []},

        deltaColorPositive : {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)},
        deltaColorNegative : {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)},

        emissive : {type : "v3", value : new THREE.Vector3(0.0, 0.0, 0.0)}
    };

    constructor(nodes : node[], settings : renderSettings)
    {
        this.nodes = nodes;

        this.initMaterial(settings);

        this.geomGen = new geometryGenerator();
        let buildRes : buildResult = this.geomGen.build(this.nodes, this.material, settings);

        this.threeMesh = buildRes.mesh;
        this.mapGeomDesc = buildRes.desc;

        this.settings = settings;
    }

    private initMaterial(settings : renderSettings) : void
    {

        if(settings.deltaColorFlipped) {
            this.setDeltaColorsFlipped();
        } else {
            this.setDeltaColorsUnflipped();
        }

        let uniforms = THREE.UniformsUtils.merge([THREE.UniformsLib["lights"], this.lightingParams]);

        let shaderCode : codeMapShaderStrings = new codeMapShaderStrings();

        this.material =  new THREE.ShaderMaterial({
            vertexShader: shaderCode.vertexShaderCode,
            fragmentShader: shaderCode.fragmentShaderCode,
            lights : true,
            uniforms : uniforms
        });
    }

    private setDeltaColorsFlipped() {
        this.lightingParams.deltaColorPositive = {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)};
        this.lightingParams.deltaColorNegative = {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)};
    }


    private setDeltaColorsUnflipped() {
        this.lightingParams.deltaColorPositive = {type : "v3", value : renderingUtil.colorToVec3(MapColors.positiveDelta)};
        this.lightingParams.deltaColorNegative = {type : "v3", value : renderingUtil.colorToVec3(MapColors.negativeDelta)};
    }

    public getThreeMesh() : THREE.Mesh {
        return this.threeMesh;
    }

    public setHighlighted(buildings : codeMapBuilding[], color? : number)
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.highlightedIndices.value = buildings.map((b : codeMapBuilding) => {return b.id;});
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numHighlights.value = buildings.length;

        if (color) {
            this.lightingParams.highlightColor.value = renderingUtil.colorToVec3(color);
        }

        this.currentlyHighlighted = buildings;
    }

    public setSelected(buildings : codeMapBuilding[], color? : number)
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.selectedIndices.value = buildings.map((b : codeMapBuilding) => {return b.id;});
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numSelections.value  = buildings.length;
        
        if (color) {
            this.lightingParams.selectedColor.value = renderingUtil.colorToVec3(color);
        }

        this.currentlySelected = buildings;
    }

    public getCurrentlyHighlighted() : codeMapBuilding[] | null
    {
        return this.currentlyHighlighted;
    }

    public getCurrentlySelected() : codeMapBuilding[] | null
    {
        return this.currentlySelected;
    }

    public clearHighlight()
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numHighlights.value = 0.0;
        this.currentlyHighlighted = null;
    }

    public clearSelected()
    {
        //noinspection TypeScriptUnresolvedVariable
        this.material.uniforms.numSelections.value = 0.0;
        this.currentlySelected = null;
    }

    public getMeshDescription() : codeMapGeometricDescription
    {
        return this.mapGeomDesc;
    }

    private calculatePickingRay(mouse : mousePos, camera : THREE.Camera) : THREE.Ray
    {

        let ray : THREE.Ray = new THREE.Ray();
        ray.origin.setFromMatrixPosition(camera.matrixWorld);
        ray.direction.set(mouse.x, mouse.y, 0.5).unproject(camera).sub(ray.origin).normalize();

        return ray;
    }

    public checkMouseRayMeshIntersection(mouse : mousePos, camera : THREE.Camera) : intersectionResult
    {
        let ray : THREE.Ray = this.calculatePickingRay(mouse, camera);
        return this.getMeshDescription().intersect(ray);
    }

    public setScale(x : number, y : number, z : number)
    {
        this.mapGeomDesc.setScales(new THREE.Vector3(x, y, z));
    }
}