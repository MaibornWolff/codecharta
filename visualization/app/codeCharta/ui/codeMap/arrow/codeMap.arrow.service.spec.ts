import { TestBed } from "@angular/core/testing"
import { StoreModule, Store, State } from "@ngrx/store"
import { CodeMapArrowService } from "./codeMap.arrow.service"
import { ThreeSceneService } from "../threeViewer/threeSceneService"
import { Object3D, Vector3 } from "three"
import {
    CODE_MAP_BUILDING,
    OUTGOING_NODE,
    DIFFERENT_NODE,
    INCOMING_NODE,
    VALID_EDGES_DECORATED,
    CODE_MAP_BUILDING_WITH_INCOMING_EDGE_NODE,
    CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE
} from "../../../util/dataMocks"
import { CcState, Node } from "../../../codeCharta.model"
import { ColorConverter } from "../../../util/color/colorConverter"
import { setScaling } from "../../../state/store/appSettings/scaling/scaling.actions"
import { setEdges } from "../../../state/store/fileSettings/edges/edges.actions"
import { setHeightMetric } from "../../../state/store/dynamicSettings/heightMetric/heightMetric.actions"
import { CodeMapMesh } from "../rendering/codeMapMesh"
import { toggleEdgeMetricVisible } from "../../../state/store/appSettings/isEdgeMetricVisible/isEdgeMetricVisible.actions"
import { wait } from "../../../util/testUtils/wait"
import { IdToBuildingService } from "../../../services/idToBuilding/idToBuilding.service"
import { appReducers, setStateMiddleware } from "../../../state/store/state.manager"
import { clone } from "../../../util/clone"
import { setShowOutgoingEdges } from "../../../state/store/appSettings/showEdges/outgoing/showOutgoingEdges.actions"
import { setShowIncomingEdges } from "../../../state/store/appSettings/showEdges/incoming/showIncomingEdges.actions"

describe("CodeMapArrowService", () => {
    let codeMapArrowService: CodeMapArrowService
    let threeSceneService: ThreeSceneService
    let store: Store<CcState>
    let state: State<CcState>
    let idToBuildingService: IdToBuildingService

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [StoreModule.forRoot(appReducers, { metaReducers: [setStateMiddleware] })]
        })
        threeSceneService = TestBed.inject(ThreeSceneService)
        store = TestBed.inject(Store)
        state = TestBed.inject(State)
        idToBuildingService = TestBed.inject(IdToBuildingService)
        codeMapArrowService = new CodeMapArrowService(store, state, threeSceneService, idToBuildingService)
    })

    function withMockedThreeSceneService() {
        threeSceneService = codeMapArrowService["threeSceneService"] = jest.fn().mockReturnValue({
            edgeArrows: {
                children: [],
                add: jest.fn()
            },
            highlightedBuildings: [],
            getMapMesh: jest.fn().mockReturnValue({
                getMeshDescription: jest.fn().mockReturnValue({ getBuildingByPath: jest.fn() }),
                clearHighlight: jest.fn(),
                highlightBuilding: jest.fn(),
                clearSelection: jest.fn(),
                selectBuilding: jest.fn()
            }),
            highlightBuildings: jest.fn(),
            addBuildingToHighlightingList: jest.fn(),
            getSelectedBuilding: jest.fn().mockReturnValue({
                value: "value"
            })
        })()
    }

    function setupEdgeArrowsWithChildren() {
        const dummyObject3D = new Object3D()
        threeSceneService.edgeArrows.children = [dummyObject3D, dummyObject3D]
    }

    function setupArrows() {
        const dummyObject3D = new Object3D()
        codeMapArrowService["arrows"] = [dummyObject3D, dummyObject3D]
    }

    describe("constructor", () => {
        it("should assign arrows an empty array", () => {
            expect(codeMapArrowService["arrows"].length).toBe(0)
        })
    })

    describe("Arrow Behaviour when selecting and hovering a building", () => {
        it("should only highlight small leaf when big leaf is selected", async () => {
            store.dispatch(setEdges({ value: VALID_EDGES_DECORATED }))
            const nodes: Node[] = [
                CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE.node,
                CODE_MAP_BUILDING_WITH_INCOMING_EDGE_NODE.node,
                DIFFERENT_NODE
            ]
            threeSceneService["mapMesh"] = new CodeMapMesh(nodes, state.getValue(), false)
            codeMapArrowService.addEdgeMapBasedOnNodes(nodes)
            codeMapArrowService.addEdgePreview()
            store.dispatch(setHeightMetric({ value: "mcc" }))

            threeSceneService.selectBuilding(CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE)
            codeMapArrowService.onBuildingHovered(CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE)

            await wait(codeMapArrowService["HIGHLIGHT_BUILDING_DELAY"])

            expect(threeSceneService["highlighted"]).toMatchSnapshot()
            expect(threeSceneService["selected"]).toMatchSnapshot()
        })
        it("should restore to previous color if another building is selected in delta mode", async () => {
            const nodes: Node[] = [
                CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE.node,
                CODE_MAP_BUILDING_WITH_INCOMING_EDGE_NODE.node,
                DIFFERENT_NODE
            ]
            threeSceneService["mapMesh"] = new CodeMapMesh(nodes, state.getValue(), true)

            store.dispatch(setHeightMetric({ value: "mcc" }))

            threeSceneService.selectBuilding(CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE)
            threeSceneService.selectBuilding(CODE_MAP_BUILDING_WITH_INCOMING_EDGE_NODE)

            expect(threeSceneService["selected"]).toMatchSnapshot()
            expect(threeSceneService["mapMesh"].getBuildingByPath(CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE.node.path)).toMatchSnapshot()
        })
        it("should debounce the edge reset of buildings to improve performance", async () => {
            const resetEdgesOfBuildingMock = jest.fn()
            codeMapArrowService["resetEdgesOfBuildings"] = resetEdgesOfBuildingMock
            codeMapArrowService.onBuildingHovered(CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE)

            expect(resetEdgesOfBuildingMock).not.toHaveBeenCalled()

            await wait(codeMapArrowService["HIGHLIGHT_BUILDING_DELAY"])

            expect(resetEdgesOfBuildingMock).toHaveBeenCalled()
        })
        it("should only add outgoing edges when showOutgoingEdges is true and node has originNode path", () => {
            withMockedThreeSceneService()
            codeMapArrowService.addArrow = jest.fn()

            const outgoingNode: Node = OUTGOING_NODE
            const incomingNode: Node = INCOMING_NODE
            const nodesMap = new Map<string, Node>()
            nodesMap.set(outgoingNode.path, outgoingNode)
            nodesMap.set(incomingNode.path, incomingNode)
            codeMapArrowService["map"] = nodesMap

            store.dispatch(setShowOutgoingEdges({ value: true }))
            store.dispatch(setShowIncomingEdges({ value: false }))
            store.dispatch(setEdges({ value: [{ fromNodeName: outgoingNode.path, toNodeName: incomingNode.path, attributes: {} }] }))

            codeMapArrowService["buildPairingEdges"](nodesMap)

            expect(codeMapArrowService.addArrow).toHaveBeenCalledTimes(1)
            expect(codeMapArrowService.addArrow).toHaveBeenCalledWith(outgoingNode, incomingNode, true)
        })

        it("should only add incoming edges when showIncomingEdges is true and node has targetNode path", () => {
            withMockedThreeSceneService()
            codeMapArrowService.addArrow = jest.fn()

            const outgoingNode: Node = OUTGOING_NODE
            const incomingNode: Node = INCOMING_NODE
            const nodesMap = new Map<string, Node>()
            nodesMap.set(outgoingNode.path, outgoingNode)
            nodesMap.set(incomingNode.path, incomingNode)
            codeMapArrowService["map"] = nodesMap

            store.dispatch(setShowOutgoingEdges({ value: false }))
            store.dispatch(setShowIncomingEdges({ value: true }))
            store.dispatch(setEdges({ value: [{ fromNodeName: outgoingNode.path, toNodeName: incomingNode.path, attributes: {} }] }))

            codeMapArrowService["buildPairingEdges"](nodesMap)

            expect(codeMapArrowService.addArrow).toHaveBeenCalledTimes(1)
            expect(codeMapArrowService.addArrow).toHaveBeenCalledWith(outgoingNode, incomingNode, false)
        })
    })

    describe("SelectionMethods", () => {
        beforeEach(() => {
            codeMapArrowService.clearArrows = jest.fn()
            codeMapArrowService.addArrow = jest.fn()
            codeMapArrowService["showEdgesOfBuildings"] = jest.fn()
            codeMapArrowService.addEdgePreview = jest.fn()
            threeSceneService.clearHighlight = jest.fn()
            codeMapArrowService["buildPairingEdges"] = jest.fn()
            codeMapArrowService.scale = jest.fn()
        })

        it("should call clearArrows and showEdgesOfBuildings through BuildingSelected", () => {
            codeMapArrowService.onBuildingSelected({ building: CODE_MAP_BUILDING })

            expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
            expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
            expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
        })

        it("should not call sub-methods through BuildingSelected if building is undefined", () => {
            codeMapArrowService.onBuildingSelected({ building: undefined })

            expect(codeMapArrowService.clearArrows).not.toHaveBeenCalled()
            expect(codeMapArrowService["showEdgesOfBuildings"]).not.toHaveBeenCalled()
            expect(codeMapArrowService.addEdgePreview).not.toHaveBeenCalled()
            expect(codeMapArrowService.scale).toHaveBeenCalled()
        })

        it("should call clearArrows and showEdgesOfBuildings through BuildingHovered", async () => {
            codeMapArrowService.onBuildingHovered(CODE_MAP_BUILDING)

            await wait(codeMapArrowService["HIGHLIGHT_BUILDING_DELAY"])

            expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
            expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
            expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
        })

        it("should call clearArrows and showEdgesOfBuildings through BuildingUnHovered when edge metric is enabled", () => {
            codeMapArrowService.onBuildingUnhovered()

            expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
            expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
            expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
        })

        it("should call clearArrows and showEdgesOfBuildings through BuildingUnHovered when edge metric is disabled", () => {
            store.dispatch(toggleEdgeMetricVisible())
            codeMapArrowService.onBuildingUnhovered()

            expect(codeMapArrowService.clearArrows).toHaveBeenCalledTimes(0)
            expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalledTimes(0)
            expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
        })

        it("should call clearArrows and showEdgesOfBuildings through BuildingDeselected", () => {
            codeMapArrowService.onBuildingDeselected()

            expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
            expect(codeMapArrowService.addEdgePreview).toHaveBeenCalled()
            expect(codeMapArrowService.addArrow).toHaveBeenCalledTimes(0)
            expect(threeSceneService.clearHighlight).toHaveBeenCalledTimes(0)
            expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalledTimes(0)
        })
    })

    describe("clearArrows", () => {
        beforeEach(() => {
            withMockedThreeSceneService()
        })

        it("should remove all array entries of field arrows", () => {
            setupEdgeArrowsWithChildren()

            codeMapArrowService.clearArrows()

            expect(codeMapArrowService["arrows"].length).toBe(0)
        })

        it("should remove all array entries of threeSceneService.edgeArrows.children", () => {
            setupEdgeArrowsWithChildren()

            codeMapArrowService.clearArrows()

            expect(threeSceneService.edgeArrows.children.length).toBe(0)
        })
    })

    describe("addEdgePreview", () => {
        beforeEach(() => {
            codeMapArrowService["map"] = new Map<string, Node>()
            codeMapArrowService["map"].get = jest.fn(() => {
                return INCOMING_NODE
            })
            codeMapArrowService["previewMode"] = jest.fn()
        })
        it("should create an edge Preview of one", () => {
            const nodes: Node[] = [OUTGOING_NODE]

            codeMapArrowService.addEdgeMapBasedOnNodes(nodes)

            expect(codeMapArrowService["map"].size).toEqual(1)
        })
        it("should not change stored edges for Preview", () => {
            codeMapArrowService.addEdgePreview()

            expect(codeMapArrowService["map"].size).toEqual(0)
        })
        it("when targetNode is invalid then it should not call preview mode", () => {
            const invalidEdge = clone(VALID_EDGES_DECORATED)
            invalidEdge[0].toNodeName = "invalid"
            store.dispatch(setEdges({ value: invalidEdge }))
            const nodes: Node[] = [CODE_MAP_BUILDING_WITH_OUTGOING_EDGE_NODE.node]

            codeMapArrowService.addEdgeMapBasedOnNodes(nodes)
            codeMapArrowService.addEdgePreview()

            expect(codeMapArrowService["previewMode"]).not.toHaveBeenCalled()
        })
        it("when originNodeName is invalid then it should not call preview mode", () => {
            const invalidEdge = clone(VALID_EDGES_DECORATED)
            invalidEdge[0].fromNodeName = "invalid"
            store.dispatch(setEdges({ value: invalidEdge }))
            const nodes: Node[] = [CODE_MAP_BUILDING_WITH_INCOMING_EDGE_NODE.node]

            codeMapArrowService.addEdgeMapBasedOnNodes(nodes)
            codeMapArrowService.addEdgePreview()

            expect(codeMapArrowService["previewMode"]).not.toHaveBeenCalled()
        })
    })

    describe("createCurve", () => {
        it("should create a curve out of the 2 Nodes", () => {
            const originNode: Node = OUTGOING_NODE
            const targetNode: Node = INCOMING_NODE
            const curveScale = 100 * state.getValue().appSettings.edgeHeight

            const curve = codeMapArrowService["createCurve"](originNode, targetNode, curveScale)

            expect(curve).toBeDefined()
        })
    })

    describe("highlightBuilding", () => {
        it("should highlight certain buildings", () => {
            withMockedThreeSceneService()

            codeMapArrowService["highlightBuilding"](OUTGOING_NODE)

            expect(threeSceneService.getMapMesh().getMeshDescription().getBuildingByPath).toHaveBeenCalled()
            expect(threeSceneService.addBuildingToHighlightingList).toHaveBeenCalled()
        })
    })

    describe("setCurveColor", () => {
        beforeEach(() => {
            threeSceneService.edgeArrows["add"] = jest.fn()
            codeMapArrowService["arrows"].push = jest.fn()
        })
        it("should run through the function with mocked subfunctions", () => {
            const originNode: Node = OUTGOING_NODE
            const targetNode: Node = INCOMING_NODE
            const curveScale = 100 * state.getValue().appSettings.edgeHeight
            const curve = codeMapArrowService["createCurve"](originNode, targetNode, curveScale)
            const color = ColorConverter.convertHexToNumber(state.getValue().appSettings.mapColors.outgoingEdge)

            codeMapArrowService["setCurveColor"](curve, color)

            expect(threeSceneService.edgeArrows["add"]).toHaveBeenCalled()
            expect(codeMapArrowService["arrows"].push).toHaveBeenCalled()
        })
    })

    describe("scale", () => {
        it("should set the scale of all arrows to x, y and z", () => {
            setupArrows()
            store.dispatch(setScaling({ value: new Vector3(1, 2, 3) }))

            codeMapArrowService.scale()

            expect(codeMapArrowService["arrows"][0].scale.x).toBe(1)
            expect(codeMapArrowService["arrows"][0].scale.y).toBe(2)
            expect(codeMapArrowService["arrows"][0].scale.z).toBe(3)
        })
    })
})
