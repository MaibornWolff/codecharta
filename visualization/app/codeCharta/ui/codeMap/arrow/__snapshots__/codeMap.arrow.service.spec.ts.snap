// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should only highlight small leaf when big leaf is selected 1`] = `
[
  CodeMapBuilding {
    "_boundingBox": Box3 {
      "isBox3": true,
      "max": Vector3 {
        "x": 6,
        "y": 8,
        "z": 10,
      },
      "min": Vector3 {
        "x": 5,
        "y": 6,
        "z": 7,
      },
    },
    "_color": "#820E0E",
    "_defaultColor": "#820E0E",
    "_defaultDeltaColor": "#000000",
    "_deltaColor": "#000000",
    "_id": 1,
    "_node": {
      "attributes": {
        "a": 20,
        "b": 15,
        "mcc": 100,
      },
      "color": "#820E0E",
      "deltas": {
        "a": 1,
        "b": 2,
      },
      "depth": 4,
      "edgeAttributes": {
        "a": {
          "incoming": 2,
          "outgoing": 666,
        },
      },
      "flat": false,
      "height": 2,
      "heightDelta": 20,
      "id": 2,
      "incomingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "isLeaf": true,
      "length": 3,
      "link": "NO_LINK",
      "mapNodeDepth": 2,
      "markingColor": "0xFFFFFF",
      "name": "root/small leaf",
      "outgoingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "path": "/root/Parent Leaf/small leaf",
      "visible": true,
      "width": 1,
      "x0": 5,
      "y0": 7,
      "z0": 6,
    },
  },
  CodeMapBuilding {
    "_boundingBox": Box3 {
      "isBox3": true,
      "max": Vector3 {
        "x": 6,
        "y": 8,
        "z": 10,
      },
      "min": Vector3 {
        "x": 5,
        "y": 6,
        "z": 7,
      },
    },
    "_color": "#820E0E",
    "_defaultColor": "#820E0E",
    "_defaultDeltaColor": "#000000",
    "_deltaColor": "#000000",
    "_id": 1,
    "_node": {
      "attributes": {
        "a": 20,
        "b": 15,
        "mcc": 100,
      },
      "color": "#820E0E",
      "deltas": {
        "a": 1,
        "b": 2,
      },
      "depth": 4,
      "edgeAttributes": {
        "a": {
          "incoming": 2,
          "outgoing": 666,
        },
      },
      "flat": false,
      "height": 2,
      "heightDelta": 20,
      "id": 2,
      "incomingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "isLeaf": true,
      "length": 3,
      "link": "NO_LINK",
      "mapNodeDepth": 2,
      "markingColor": "0xFFFFFF",
      "name": "root/small leaf",
      "outgoingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "path": "/root/Parent Leaf/small leaf",
      "visible": true,
      "width": 1,
      "x0": 5,
      "y0": 7,
      "z0": 6,
    },
  },
]
`;

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should only highlight small leaf when big leaf is selected 2`] = `
CodeMapBuilding {
  "_boundingBox": Box3 {
    "isBox3": true,
    "max": Vector3 {
      "x": -Infinity,
      "y": -Infinity,
      "z": -Infinity,
    },
    "min": Vector3 {
      "x": Infinity,
      "y": Infinity,
      "z": Infinity,
    },
  },
  "_color": "#EB8319",
  "_defaultColor": "#ddcc00",
  "_defaultDeltaColor": "#000000",
  "_deltaColor": "#EB8319",
  "_id": 1,
  "_node": {
    "attributes": {
      "a": 20,
      "b": 15,
      "mcc": 14,
    },
    "color": "#AABBCC",
    "deltas": {
      "a": 1,
      "b": 2,
    },
    "depth": 4,
    "edgeAttributes": {
      "a": {
        "incoming": 2,
        "outgoing": 666,
      },
    },
    "flat": false,
    "height": 2,
    "heightDelta": 20,
    "id": 1,
    "incomingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "isLeaf": true,
    "length": 3,
    "link": "NO_LINK",
    "mapNodeDepth": 2,
    "markingColor": "0xFFFFFF",
    "name": "root/big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "path": "/root/big leaf",
    "visible": true,
    "width": 1,
    "x0": 5,
    "y0": 7,
    "z0": 6,
  },
}
`;

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should restore to previous color if another building is selected in delta mode 1`] = `
CodeMapBuilding {
  "_boundingBox": Box3 {
    "isBox3": true,
    "max": Vector3 {
      "x": -Infinity,
      "y": -Infinity,
      "z": -Infinity,
    },
    "min": Vector3 {
      "x": Infinity,
      "y": Infinity,
      "z": Infinity,
    },
  },
  "_color": "#EB8319",
  "_defaultColor": "#ddcc00",
  "_defaultDeltaColor": "#000000",
  "_deltaColor": "#EB8319",
  "_id": 2,
  "_node": {
    "attributes": {
      "a": 20,
      "b": 15,
      "mcc": 100,
    },
    "color": "#820E0E",
    "deltas": {
      "a": 1,
      "b": 2,
    },
    "depth": 4,
    "edgeAttributes": {
      "a": {
        "incoming": 2,
        "outgoing": 666,
      },
    },
    "flat": false,
    "height": 2,
    "heightDelta": 20,
    "id": 2,
    "incomingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "isLeaf": true,
    "length": 3,
    "link": "NO_LINK",
    "mapNodeDepth": 2,
    "markingColor": "0xFFFFFF",
    "name": "root/small leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "path": "/root/Parent Leaf/small leaf",
    "visible": true,
    "width": 1,
    "x0": 5,
    "y0": 7,
    "z0": 6,
  },
}
`;

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should restore to previous color if another building is selected in delta mode 2`] = `
CodeMapBuilding {
  "_boundingBox": Box3 {
    "isBox3": true,
    "max": Vector3 {
      "x": 6,
      "y": 8,
      "z": 10,
    },
    "min": Vector3 {
      "x": 5,
      "y": 6,
      "z": 7,
    },
  },
  "_color": "#6987A5",
  "_defaultColor": "#AABBCC",
  "_defaultDeltaColor": "#000000",
  "_deltaColor": "#1A1A1A",
  "_id": 0,
  "_node": {
    "attributes": {
      "a": 20,
      "b": 15,
      "mcc": 14,
    },
    "color": "#AABBCC",
    "deltas": {
      "a": 1,
      "b": 2,
    },
    "depth": 4,
    "edgeAttributes": {
      "a": {
        "incoming": 2,
        "outgoing": 666,
      },
    },
    "flat": false,
    "height": 2,
    "heightDelta": 20,
    "id": 1,
    "incomingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "isLeaf": true,
    "length": 3,
    "link": "NO_LINK",
    "mapNodeDepth": 2,
    "markingColor": "0xFFFFFF",
    "name": "root/big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "path": "/root/big leaf",
    "visible": true,
    "width": 1,
    "x0": 5,
    "y0": 7,
    "z0": 6,
  },
}
`;
